java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestGenerationUtils.java

```diff
@@ -152,16 +166,294 @@ public static WorkspaceEdit generateTests(List<Object> arguments, IProgressMonit
+        final StringBuilder buf = new StringBuilder();
+        if (!packageFragment.isDefaultPackage()) {
+            buf.append("package ")
+                .append(packageFragment.getElementName())
+                .append(";")
+                .append(lineDelimiter)
+                .append(lineDelimiter)
+                .append("import ")
+                .append(getTestAnnotation(testKind))
+                .append(";")
+                .append(lineDelimiter)
+                .append(lineDelimiter);
+        }
+        buf.append(typeContent);
+        return buf.toString();
+    }
+
+    private static String constructTypeStub(ICompilationUnit testUnit, List<String> methods,
+            TestKind testKind, String lineDelimiter) throws CoreException {
+        final String typeName = testUnit.getElementName().replace(".java", "");
+        final StringBuilder buf = new StringBuilder();
+        buf.append("public class ").append(typeName).append(" {").append(lineDelimiter);
+        for (final String method : methods) {
+            buf.append(constructMethodStub(testUnit, testKind, method, lineDelimiter)).append(lineDelimiter);
+        }
+        buf.append("}").append(lineDelimiter);
+        return buf.toString();
+    }
+
+    private static String constructMethodStub(ICompilationUnit testUnit, TestKind testKind,
+            String method, String lineDelimiter) {
+        final StringBuilder buf = new StringBuilder();
+        buf.append("@Test").append(lineDelimiter);
+        if (testKind == TestKind.JUnit) {
+            buf.append("public ");
+        }
+        final String methodName = getTestMethodName(method);
+        buf.append("void ").append(methodName).append("() {").append(lineDelimiter).append(lineDelimiter).append("}");
+        final String methodContent = buf.toString();
+        // TODO: get test unit options directly
+        return CodeFormatterUtil.format(CodeFormatter.K_STATEMENTS, methodContent, 1,
+                lineDelimiter, testUnit.getJavaProject().getOptions(true));
+    }
+
+    private static IClasspathEntry getTestClasspathEntry(ICompilationUnit unit) throws JavaModelException {
+        final IJavaProject javaProject = unit.getJavaProject();
+        // In most cases, this is the classpath entry used for testing, we first find the target entry by hard-code
+        // to avoid go into the generated entries.
+        final IClasspathEntry testEntry = javaProject.getClasspathEntryFor(
+            javaProject.getPath().append("src/test/java"));
+        if (ProjectTestUtils.isTestEntry(testEntry)) {
+            return testEntry;
+        }
+
+        final IClasspathEntry[] entries = javaProject.readRawClasspath();
+        for (final IClasspathEntry entry : entries) {
+            if (ProjectTestUtils.isTestEntry(entry)) {
+                return entry;
+            }
+        }
+
+        return entries[0];
```

Comment:
If no test folder, disable this source action or generate test in the same place as source file.

Reply:
How about generate test under the same classpath entry with the source file.

Reply:
that's ok. You can even create a test classpath entry.